Some refactors around tests, randomness and phase2

wzmuda opened the pull request:
When working on the online mode, I started with refactoring of the existing codebase:
- I changed the structure of integration tests, so it will be easier to plug another test suite into it.
- I refactored the randomness package, so that it is mockable in tests. This way tests don't require internet connection now.
- I refactored the phase2 package a little, by moving the timestamping of the contribution files to the command line handler. This is because timestamping is only needed in the CLI mode but will not be necessary in the online mode.

Please see individual commit logs for details.

wzmuda committed:
test: modularize, rename and comment

Refator the integration test to make room for client/server test.

The integration test is renamed to reflect its testing the offline mode
of the program. The circuit definition is moved to another file, so it's
clearly separated from the test logic. It will be later used by the
online mode test. Separate test resources from code.

Signed-off-by: Wojciech Zmuda <zmuda.w@gmail.com>

wzmuda committed:
phase2: contribute: accept output path as argument

Modify Contribute() to accept both input and output Phase 2 file
path instead of just input path. Previously the function generated the
output path by appending timestamp to the input path. Now Contribute()
just generates contribution to the given file.

Move the path generation duty to the CLI command handler. The program
still works in the same way from the user perspective.

This change will allow to add the online mode easier, because
timestamping contributions will not be necessary in that case.

Signed-off-by: Wojciech Zmuda <zmuda.w@gmail.com>

wzmuda committed:
randomness: make testable, mockable and add tests

Refactor the package so it has a settable provider. Convert the existing
init function to the real drand-based provided. Add another mock
provider that returns a fixed value instead of a proper random one.

Override the drand provided with the mock provider in tests. This way
tests can run and pass when e.g. the service is dead or when I'm on the
plane again with no internet access.

Move it to utils because this is where all the goodies that are helpful
but not really related to the trusted setup itself are going to live.

Signed-off-by: Wojciech Zmuda <zmuda.w@gmail.com>

batmendbar commented on phase2/phase2.go:
```diff
@@ -78,20 +58,19 @@ func updateTimestamp(str string) string {
 //
 // Returns the file name of the Phase 2 object containing the contributions and nil on success and empty string
```
Since we longer returns the file name, I think we should change this comment

wzmuda replied:
Good catch, fixed

batmendbar reviewed (approved):
One very minor issue with an outdated comment. Otherwise, looks good to me!

piohei commented on utils/randomness/randomness.go:
```diff
@@ -0,0 +1,35 @@
+}
+
+// global provider used by package functions
+var provider BeaconProvider
```
I'm not a big fun of singleton as it is hard sometimes to test things in parallel later (in some cases). But here it works so just sharing thoughts. ;)

wzmuda replied:
You're right but not because of singleton :D I was too focused on the not-yet-existing online part, where you would fire up the server and it would generate this random beacon on the init. The same value is needed through the whole life of the program, so the pattern I implemented here makes sense (is it a singleton though? It just initializes on the start vs on demand, because the demand happens once, at the start).

And this totally breaks the offline mode, where user orchestrates the ceremony manually. In the offline mode there are multiple runs of the program. When you run `./trusted-setup init` it will initialize the ceremony with some random value, but then when you call `./trusted-setup extract-keys`, it will most probably generate another random value (which will break the process).

This problem was not introduced in this PR, but was there from the start. I did not catch this in tests, because they do everything in one go, so the random value does not change.

wzmuda replied:
I fixed this in two ways:
1. I added a new commit that fixes the issue I described above: https://github.com/reilabs/trusted-setup/pull/5/commits/bff539978f96a51a05e1cfed77a98f566d8406c4
2. The randomness refactoring commit does not get the randomness at init anymore. It is still used like this by the program, but it is not the feature of the randomness package anymore, i.e. if we want in the future, we can query GetBeacon() and we will get whatever drand provides at any given moment.

`var provider BeaconProvider` stays but this is not what makes this package a singleton. It's just a handler for randomness provider and we have two - the acual one and the mock one, which will be more helpful in the online cermony tests (because we cannot easily inject a well-known random value there).

piohei replied:
In most cases when you run a tests you run them as a single program. On that program start `init()` is called and `provider` is set. Then when tests are run in parallel few threads are being spawn (each for test execution). Let's imagine that you have two tests. So you have two tests running same time that will control single `provider` value set in module. If test one want to mock provider to return for example value A, an test two want to mock provider to return value B, with current approach it is impossible or creates race condition. You can be sure that during run of test one which expects value A test two wouldn't change provider so it will return value B. Than assert in test one will brake as A != B.

wzmuda replied:
Ok, so now the package has two randomness providers, both implementing the same interface. One uses drand, the other is mock. They're absolutely independent, no state is preserved, you can instantiate multiple mocks with different values and you can query drand multiple times and you'll get whatever the backend returns.

Now for the offline ceremony it's trivial to inject randomness (actually you don't need the mock provider for that), but for the online ceremony the coordinator module will accept the randomness interface and in tests a mock will be passed to its constructor.

wzmuda committed:
phase2: accept beacon as argument

The randomness package generates beacon once at init. This does not make
sense in the offline mode, where the application is ran multiple times.
Such approach will generate a different random value at each run, while
it must be the same at init and at extract-keys.

Change the approach. The extract-keys command now accepts a --beacon
argument. User is expect to pass the same beacon that was generated and
printed on the init command.

Signed-off-by: Wojciech Zmuda <zmuda.w@gmail.com>
